src/wrapper/ddprof_wrapper.cc

```diff
@@ -0,0 +1,159 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0. This product includes software
+// developed at Datadog (https://www.datadoghq.com/). Copyright 2021-Present
+// Datadog, Inc.
+
+#include "ddprof_wrapper.h"
+
+extern "C" {
+#include "ddprof_cmdline.h"
+}
+#include "defer.hpp"
+
+#include <cerrno>
+#include <chrono>
+#include <cstdio>
+#include <cstdlib>
+#include <fcntl.h>
+#include <sys/mman.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <thread>
+#include <unistd.h>
+
+extern const char _binary_ddprof_start[];
+extern const char _binary_ddprof_end[];
+
+namespace {
+struct ProfilerState {
+  bool started = false;
+  pid_t profiler_pid = 0;
+};
+
+ProfilerState g_state;
+
+struct ProfilerAutoStart {
+  ProfilerAutoStart() {
+    const char *autostart_env = getenv("DD_PROFILING_NATIVE_AUTOSTART");
+    if (autostart_env && arg_yesno(autostart_env, 1)) {
+      ddprof_start_profiling();
+    }
+  }
+
+  ~ProfilerAutoStart() {
+    if (g_state.started) {
+      ddprof_stop_profiling(1000);
+    }
+  }
+};
+
+ProfilerAutoStart g_autostart;
+} // namespace
+
+int ddprof_start_profiling() {
+  if (g_state.started) {
+    return -1;
+  }
+  pid_t target_pid = getpid();
+  int pipefd[2];
+  if (pipe2(pipefd, O_CLOEXEC) == -1) {
+    return -1;
+  }
+
+  pid_t middle_pid = fork();
+  if (middle_pid == -1) {
+    return -1;
+  }
+
+  if (middle_pid == 0) {
+    // executed in middle
+    close(pipefd[0]);
+
+    auto defer_pipeclose = make_defer([&pipefd] { close(pipefd[1]); });
+
+    middle_pid = getpid();
+    if (middle_pid == -1) {
+      return -1;
+    }
+
+    pid_t grandchild_pid = fork();
+    if (grandchild_pid == 0) {
+      // executed in grand child
+      grandchild_pid = getpid();
+      int fd = memfd_create("ddprof", MFD_CLOEXEC);
+      if (fd == -1) {
+        return -1;
+      }
+      defer { close(fd); };
+
+      if (write(fd, _binary_ddprof_start,
+                _binary_ddprof_end - _binary_ddprof_start) == -1) {
+        return -1;
+      }
+
+      char path_buf[32];
+      snprintf(path_buf, sizeof(path_buf), "/proc/self/fd/%d", fd);
```

Comment:
Tricky!  The `memfd()` anonymous regions are backed by tmpfs, but referenced on the VFS only through procfs--I think this is probably the only way to get a path reference to once (which is weird, because the procfs fd entry should be a symlink, haha).

Reply:
Yes, it's not very clean, I could use `fexecve` but it does not offer the inheritance of environment as `execl`. But perhaps I could pass `environ` directly to `fexecve` ?

Reply:
And for the record, I personally think this is pretty close to as clean as it could be.  The interfaces are rather limited.  When I said it was tricky, I meant that the idea itself has several very subtle implications, like the fact that you're making a memfd and then executing it, but there's no conspicuous canonical path other than in `/proc`.  I think this is great!